nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/internal/metadata/ArtifactsXmlAbsoluteUrlRemover.java

```diff
@@ -68,66 +70,150 @@
+        //try-with-resources will be better here, but XMLEventReader and XMLEventWriter are not AutoCloseable
+        try {
+          reader = inputFactory.createXMLEventReader(xmlIn);
+          writer = outputFactory.createXMLEventWriter(xmlOut);
+          transformer.transform(reader, writer);
+          writer.flush();
+        }
+        finally {
+          if (reader != null) {
+            reader.close();
+          }
+          if (writer != null) {
+            writer.close();
           }
-          return convertFileToTempBlob(tempFile, repository);
         }
       }
-      finally {
-        delete(tempFile);
-        delete(artifactsTempFile);
+      catch (XMLStreamException ex) {
+        log.error("Failed to rewrite metadata for file with extension {} and blob {} with reason: {} ",
+            ex, artifact.getBlob().getId(), ex);
+        return artifact;
       }
+      return convertFileToTempBlob(tempFile, repository);
     }
-    catch (IOException | XMLStreamException ex) {
-      log.error("Failed to fix absolute urls for file with extension {} and blob {} with reason: {} ",
-          ex, artifact.getBlob().getId(), ex);
-      return artifact;
+    finally {
+      delete(tempFile);
+      delete(artifactsTempFile);
     }
   }
 
-  private InputStream xmlInputStream(final TempBlob artifact, final String extension, final Path artifactsTempFile)
+  private void changeLocationToAbsoluteInCompositeRepository(
+      final XMLEventReader reader,
+      final XMLEventWriter writer,
+      final URI remoteUrl,
+      final String nexusRepositoryName) throws XMLStreamException
+  {
+    XMLEvent previous = null;
+
+    List<XMLEvent> buffer = new ArrayList<>();
+
+    while (reader.hasNext()) {
+      XMLEvent event = reader.nextEvent();
+
+      if ((isStartTagWithName(previous, "children") || isEndTagWithName(previous, "child")) &&
+          isStartTagWithName(event, "child")) {
+        buffer.add(changeLocationAttribute((StartElement) event, remoteUrl, nexusRepositoryName));
+      }
+      else {
+        buffer.add(event);
+      }
+
+      for (XMLEvent xmlEvent : buffer) {
+        writer.add(xmlEvent);
+      }
+      buffer.clear();
+
+      if (!event.toString().trim().isEmpty()) {
+        previous = event;
+      }
+    }
+  }
+
+  private XMLEvent changeLocationAttribute(
+      final StartElement event,
+      final URI remoteUrl,
+      final String nexusRepositoryName)
+  {
+    QName location = new QName("location");
+
+    String locationValue;
+    String value = event.getAttributeByName(location).getValue();
+    URI uri = URI.create(value);
+    if (uri.isAbsolute()) {
+      locationValue = P2PathUtils.DIVIDER + REPOSITORY + P2PathUtils.DIVIDER + nexusRepositoryName + P2PathUtils.DIVIDER + value;
+    }
+    else {
+      locationValue = P2PathUtils.DIVIDER + REPOSITORY + P2PathUtils.DIVIDER + nexusRepositoryName + P2PathUtils.DIVIDER + remoteUrl.resolve(URI.create(value)).toString();
+    }
```

Comment:
may be:
String assetPath = locationUri.isAbsolute(locationUri)? locationUri.toString(): remoteUrl.resolve(locationUri).toString() 
String locationValue = String.Format("\/repository\/%s\/%s", nexusRepositoryName, assetPath); (move the string literal to constants)

Reply:
done